vpr/src/place/analytical_placement/PlacementLegalizer.cpp

```diff
@@ -2,106 +2,541 @@
+    //      - an unplaceable atom is one that cannot fit into the real sub-tiles.
+    // TODO: Maybe combine these so that these become contiguous (N). -VB
+    std::vector<PlaceSubTile> sub_tiles;
+    PlaceSubTile overflow_sub_tile;
+    // TODO: It may be a good idea to use a strong ID for the sub_tile.
+    static constexpr size_t overflow_sub_tile_idx = std::numeric_limits<size_t>::max();
+    // Neighbors of this tile (i.e. tiles directly adjacent to this tile).
+    //  - These are other tiles that directly share a side with this tile.
+    //  - Does not include diagonals (TODO: should it?)
+    // FIXME: This should probably be left to the architecture model.
+    std::vector<PlaceTileId> neighbors;
+
+    PlaceTile(int tile_x, int tile_y, int tile_layer_num, size_t index,
+              const t_physical_tile_type &tile_type)
+                : tile_index(index), physical_tile_type(tile_type), x(tile_x),
+                  y(tile_y), layer_num(tile_layer_num) {
+        sub_tiles.resize(tile_type.capacity);
+        // Safety check; We use the largest representable unsigned integer
+        // as the overflow_sub_tile_idx. Therefore, we need to make sure
+        // the number of sub_tiles never gets that large!
+        // TODO: Make this a weaker assert level
+        VTR_ASSERT(sub_tiles.size() <= overflow_sub_tile_idx);
+    }
 
-static inline size_t getBinId(size_t x, size_t y) {
-    size_t grid_width = g_vpr_ctx.device().grid.width();
-    return x + (y * grid_width);
-}
+    // Static method to detect if the sub_tile_idx is representing the overflow
+    // sub_tile.
+    static inline bool is_overflow_sub_tile(size_t sub_tile_idx) {
+        return sub_tile_idx == overflow_sub_tile_idx;
+    }
 
-static inline void getNeighborsOfBin(const Bin& bin, std::vector<size_t>& neighbors) {
-    size_t grid_width = g_vpr_ctx.device().grid.width();
-    size_t grid_height = g_vpr_ctx.device().grid.width();
-    if (bin.x_pos >= 1)
-        neighbors.push_back(getBinId(bin.x_pos - 1, bin.y_pos));
-    if (bin.x_pos <= grid_width - 2)
-        neighbors.push_back(getBinId(bin.x_pos + 1, bin.y_pos));
-    if (bin.y_pos >= 1)
-        neighbors.push_back(getBinId(bin.x_pos, bin.y_pos - 1));
-    if (bin.y_pos <= grid_height - 2)
-        neighbors.push_back(getBinId(bin.x_pos, bin.y_pos + 1));
-}
+    // Get the sub_tile at the given index.
+    std::reference_wrapper<PlaceSubTile> get_sub_tile(size_t sub_tile_idx) {
+        // If this is the overflow index, return the overflow tile.
+        if (is_overflow_sub_tile(sub_tile_idx))
+            return overflow_sub_tile;
+        // TODO: Definately make these ones debug.
+        VTR_ASSERT(!sub_tiles.empty());
+        VTR_ASSERT(sub_tile_idx < sub_tiles.size());
+        // Else, return the subtile.
+        return sub_tiles[sub_tile_idx];
+    }
+
+    // Add an atom to a given sub_tile.
+    void add_atom_to_sub_tile(size_t node_id, size_t sub_tile_idx) {
```

Comment:
Should these two API support a way for caller to know if the atom exists already in the `add` function, or it does not exist in the `remove` function?